Mypy detects several possible type errors

Since the projects has several type annotations I decided to run mypy on it to see what it reports.

```
zyga@superfx:~/git/dco-check$ mypy .
setup.py:15: error: No library stub file for module 'setuptools'
setup.py:15: note: (Stub files are from https://github.com/python/typeshed)
dco_check/dco_check.py:288: error: Incompatible return value type (got "Optional[Match[str]]", expected "bool")
dco_check/dco_check.py:568: error: Argument 1 to "split_commits_data" has incompatible type "Optional[str]"; expected "str"
dco_check/dco_check.py:587: error: Argument 4 to "CommitInfo" has incompatible type "Optional[str]"; expected "str"
dco_check/dco_check.py:588: error: Argument 5 to "CommitInfo" has incompatible type "Optional[str]"; expected "str"
dco_check/dco_check.py:661: error: Argument 1 to "fetch_branch" has incompatible type "Optional[str]"; expected "str"
dco_check/dco_check.py:906: error: Unsupported operand types for + ("str" and "None")
dco_check/dco_check.py:906: note: Right operand is of type "Optional[str]"
dco_check/dco_check.py:913: error: "pprint" does not return a value
dco_check/dco_check.py:990: error: 'None' object is not iterable
dco_check/dco_check.py:1065: error: Item "None" of "Optional[CommitDataRetriever]" has no attribute "name"
dco_check/dco_check.py:1077: error: Item "None" of "Optional[CommitDataRetriever]" has no attribute "get_commit_range"
dco_check/dco_check.py:1092: error: Item "None" of "Optional[CommitDataRetriever]" has no attribute "get_commits"
Found 12 errors in 2 files (checked 3 source files)
```

It seems most of those can be reduced to missing None checks.

--- a/.circleci/config.yml
+++ b/.circleci/config.yml
@@ -16,6 +16,7 @@ jobs:
           command: |
             pip install -U pytest pytest-cov pytest-repeat
             pip install -U flake8 flake8-blind-except flake8-builtins flake8-class-newline flake8-comprehensions flake8-deprecated flake8-docstrings flake8-import-order flake8-quotes pep8-naming pylint
+            pip install -U mypy
       - run:
           name: Run tests
           command: python -m pytest --cov=dco_check --cov-branch --cov-report=xml

--- a/.github/workflows/test.yml
+++ b/.github/workflows/test.yml
@@ -22,6 +22,7 @@ jobs:
       run: |
         pip install -U pytest pytest-cov pytest-repeat
         pip install -U flake8 flake8-blind-except flake8-builtins flake8-class-newline flake8-comprehensions flake8-deprecated flake8-docstrings flake8-import-order flake8-quotes pep8-naming pylint
+        pip install -U mypy
     - name: Run tests
       run: |
         python -m pytest --cov=dco_check --cov-branch --cov-report=xml

--- a/.gitlab-ci.yml
+++ b/.gitlab-ci.yml
@@ -10,6 +10,7 @@ Test:
   before_script:
     - pip install -U pytest pytest-cov pytest-repeat
     - pip install -U flake8 flake8-blind-except flake8-builtins flake8-class-newline flake8-comprehensions flake8-deprecated flake8-docstrings flake8-import-order flake8-quotes pep8-naming pylint
+    - pip install -U mypy
     - pip install -U codecov
   script:
     - python3 -m pytest --cov=dco_check --cov-branch --cov-report=xml

--- a/.travis.yml
+++ b/.travis.yml
@@ -11,6 +11,7 @@ dist: bionic
 install:
   - pip install -U pytest pytest-cov pytest-repeat
   - pip install -U flake8 flake8-blind-except flake8-builtins flake8-class-newline flake8-comprehensions flake8-deprecated flake8-docstrings flake8-import-order flake8-quotes pep8-naming pylint
+  - pip install -U mypy
   - pip install -U codecov
 script:
   - python3 -m pytest --cov=dco_check --cov-branch --cov-report=xml

--- a/appveyor.yml
+++ b/appveyor.yml
@@ -12,6 +12,7 @@ environment:
 install:
   - "%PYTHON%\\python.exe -m pip install -U pytest pytest-cov pytest-repeat"
   - "%PYTHON%\\python.exe -m pip install -U flake8 flake8-blind-except flake8-builtins flake8-class-newline flake8-comprehensions flake8-deprecated flake8-docstrings flake8-import-order flake8-quotes pep8-naming pylint"
+  - "%PYTHON%\\python.exe -m pip install -U mypy"
   - "%PYTHON%\\python.exe -m pip install -U codecov"
 build: off
 test_script:

--- a/azure-pipelines.yml
+++ b/azure-pipelines.yml
@@ -20,6 +20,7 @@ jobs:
   - script: |
       pip install -U pytest pytest-cov pytest-repeat
       pip install -U flake8 flake8-blind-except flake8-builtins flake8-class-newline flake8-comprehensions flake8-deprecated flake8-docstrings flake8-import-order flake8-quotes pep8-naming pylint
+      pip install -U mypy
       pip install -U codecov
     displayName: 'Install test dependencies'
   - script: |
@@ -52,6 +53,7 @@ jobs:
   - script: |
       pip install -U pytest pytest-cov pytest-repeat
       pip install -U flake8 flake8-blind-except flake8-builtins flake8-class-newline flake8-comprehensions flake8-deprecated flake8-docstrings flake8-import-order flake8-quotes pep8-naming pylint
+      pip install -U mypy
       pip install -U codecov
     displayName: 'Install test dependencies'
   - script: |

--- a/dco_check/dco_check.py
+++ b/dco_check/dco_check.py
@@ -449,14 +449,14 @@ def extract_name_and_email(
 
 
 def format_name_and_email(
-    name: str,
-    email: str,
+    name: Optional[str],
+    email: Optional[str],
 ) -> str:
     """
     Format a name and a email into a 'name <email>' string.
 
-    :param name: the name
-    :param email: the email
+    :param name: the name, or `None` if N/A
+    :param email: the email, or `None` if N/A
     :return: the formatted string
     """
     return f"{name or 'N/A'} <{email or 'N/A'}>"
@@ -497,8 +497,8 @@ class CommitInfo:
         commit_hash: str,
         title: str,
         body: List[str],
-        author_name: str,
-        author_email: str,
+        author_name: Optional[str],
+        author_email: Optional[str],
         is_merge_commit: bool = False,
     ) -> None:
         """Create a CommitInfo object."""
@@ -572,7 +572,7 @@ class GitRetriever(CommitDataRetriever):
     ) -> Optional[List[CommitInfo]]:  # noqa: D102
         ignore_merge_commits = not check_merge_commits
         commits_data = get_commits_data(base, head, ignore_merge_commits=ignore_merge_commits)
-        commits = []
+        commits: List[CommitInfo] = []
         if commits_data is None:
             return commits
         individual_commits = split_commits_data(commits_data)
@@ -663,6 +663,8 @@ class GitlabRetriever(GitRetriever):
                 return None
             return target_branch_sha, commit_hash_head
         else:
+            if not default_branch:
+                return None
             # Otherwise test all commits off of the default branch
             logger.verbose_print(
                 f"\ton branch '{current_branch}': "
@@ -915,14 +917,14 @@ class GitHubRetriever(CommitDataRetriever):
         connection = http.client.HTTPSConnection('api.github.com')
         headers = {
             'User-Agent': 'dco_check',
-            'Authorization': 'token ' + self.github_token,
+            'Authorization': 'token ' + (self.github_token or ''),
         }
         connection.request('GET', compare_url, headers=headers)
         response = connection.getresponse()
         if 200 != response.getcode():  # pragma: no cover
-            from pprint import pprint
+            from pprint import pformat
             logger.print('Request failed: compare_url')
-            logger.print('reponse:', pprint(response.read().decode()))
+            logger.print('reponse:', pformat(response.read().decode()))
             return None
 
         # Extract data
@@ -1077,6 +1079,9 @@ def main(argv: Optional[List[str]] = None) -> int:
         if retriever.applies():
             commit_retriever = retriever
             break
+    if not commit_retriever:
+        logger.print('Could not find an applicable GitRetriever')
+        return 1
     logger.print('Detected:', commit_retriever.name())
 
     # Get default branch from remote if enabled

--- a/setup.cfg
+++ b/setup.cfg
@@ -39,6 +39,7 @@ tests_require =
     flake8-docstrings
     flake8-import-order
     flake8-quotes
+    mypy
     pep8-naming
     pylint
     pytest

--- a/test/test_mypy.py
+++ b/test/test_mypy.py
@@ -1,0 +1,32 @@
+# Copyright 2021 Christophe Bedard
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+from pathlib import Path
+import sys
+
+from mypy import api
+
+
+def test_mypy():
+    # Make sure to only check the source
+    source_dir = os.path.join(str(Path(__file__).parents[1]), 'dco_check')
+    normal_report, error_report, exit_status = api.run([source_dir])
+
+    if normal_report:
+        print(normal_report)
+    if error_report:
+        print(error_report, file=sys.stderr)
+
+    assert 0 == exit_status, 'mypy found errors'

